python/tvm/runtime/graph_runtime_factory.py

```diff
@@ -0,0 +1,144 @@
+    graph_json_str : str or graph class
+        The graph to be deployed in json format output by nnvm graph.
+        The graph can only contain one operator(tvm_op) that
+        points to the name of PackedFunc in the libmod.
+    libmod : tvm.Module
+        The module of the corresponding function
+    Returns
+    -------
+    graph_module : GraphModule
+        Runtime graph module that can be used to execute the graph.
+    """
+    if not isinstance(graph_json_str, string_types):
+        try:
+            graph_json_str = graph_json_str._tvm_graph_json()
+        except AttributeError:
+            raise ValueError("Type %s is not supported" % type(graph_json_str))
+    fcreate = get_global_func("tvm.graph_runtime_factory.create")
+    args = []
+    for k, v in params.items():
+        args.append(k)
+        args.append(ndarray.array(v))
+    return GraphRuntimeFactoryModule(fcreate(graph_runtime_kind, graph_json_str, libmod, module_name, *args))
+
+
+class GraphRuntimeFactoryModule(Module):
+    """Graph runtime factory module.
+
+    This is a module of graph runtime factory
+
+    Parameters
+    ----------
+    module : Module
+        The interal tvm module that holds the actual graph functions.
+
+    Attributes
+    ----------
+    module : Module
+        The interal tvm module that holds the actual graph functions.
+    """
+
+    def __init__(self, module):
+        self.module = module
+        self._select_module = module["select_module"]
+        self._import_module = module["import_module"]
+        self.selected_module = None
+        self.graph_json = None
+        self.lib = None
+        self.params = {}
+        self.iter_cnt = 0
+        super(GraphRuntimeFactoryModule, self).__init__(self.module.handle)
+
+    def __del__(self):
+        pass
+
+    def runtime_create(self, ctx):
+        """Create the runtime using ctx
+
+        Parameters
+        ----------
+        ctx : TVMContext or list of TVMContext
+        """
+        ctx, num_rpc_ctx, device_type_id = get_device_ctx(self.selected_module, ctx)
+        if num_rpc_ctx == len(ctx):
+            fcreate = ctx[0]._rpc_sess.get_function("tvm.graph_runtime_factory.runtime_create")
+        else:
+            fcreate = get_global_func("tvm.graph_runtime_factory.runtime_create")
+        return fcreate(self.selected_module, *device_type_id)
+
+    def import_module(self, mod):
+        """Create the runtime using ctx
+
+        Parameters
+        ----------
+        mod : GraphRuntimeFactoryModule
+            The graph runtime factory module we want to import
+        """
+        return self._import_module(mod)
+
+    def __getitem__(self, key='default'):
+        """Get specific module
+
+        Parameters
+        ----------
+        key : str
+            The key of module.
+        """
+        self.selected_module = self._select_module(key)
+        self.selected_module._entry = self.runtime_create
```

Comment:
I am not sure how this "_entry" is being used . But this kind of logic always run into race condition bug, and it is very hard to debug at later point of time.
May be if we want to have a utility function associated to each module, we can have an independent one, rather than one associated with the instance.

Reply:
`_entry` is one attribute of `module`, which will be called when we call `__call__`. For example, we have this:
```python
    with relay.build_config(opt_level=3):
        complied_graph_lib = relay.build_module.build(mod, "llvm", params=params)
    # raw api
    ctx = tvm.cpu()
    gmod = complied_graph_lib['default'](ctx)
```

When we handle `complied_graph_lib['default']`, we will call `select_module` of `GraphRuntimeFactoryModule`, then we have `complied_graph_lib['default'](ctx)`, we will enter the `__call__` of `GraphRuntimeFactoryModule` (i.e. the `__call__` of `module, see here: https://github.com/apache/incubator-tvm/blob/master/python/tvm/runtime/module.py#L106). The `_entry` could make us call the right function of `runtime.create`.

Reply:
@FrozenGene : Thanks a lot for explaining in detail! I got your idea now clearly. My worry is the func attached to "_entry" has dependency on "selected_module" attribute. As "selected_module" may have possibly changed ? 
Please correct me if my understanding is wrong.

Reply:
This will have no problem. Because the `_entry` is a functor ( `runtime_create`). Even the `selected_module` is changed, we will share the same `runtime_create`. Please the example of multi models example (https://github.com/apache/incubator-tvm/blob/784a6bdbecc0ff3040d003c90ede5794b8be816a/tests/python/unittest/test_module_runtime_interface.py#L95).

Reply:
Yes, you are right! 
But if selected_module is changed. Then get_device_ctx() inside runtime_create() may yield unintended results?

Reply:
Not. Because when we call runtime create, we will pass the ctx argument to it. You could see the test case I list in previous comment. Inside that case, we have two models but with different ctx (cpu and gpu)

Reply:
Sorry for such late response! I went through the test cases, it is perfectly fine. However the scenario i am talking about is not a usual one.  
I understand that the standard syntax you have followed for fetching each module is as below:
`gmod = complied_graph_lib['mod_name'](ctx)`

By using above syntax, there is no issue, as the dependency of "selected_module" in "runtime_create" is over.

But as it is not a strict syntax, user can also use it as below or some other way:
```
mod1 = complied_graph_lib['mod1_name']
mod2 = complied_graph_lib['mod2_name']

gmod1 = mod1(ctx1)
gmod2 = mod2(ctx2)

```
I have verified above scenario by modifying `test_multi_models()`, it indeed raise issue.
My point here is, there is a chance it might cause mishap, at any point of time. So may be either, we can make the syntax strict or we can remove the dependency of the `runtime_create` on `selected_module`.

It is just a suggestion to avoid any future issues :)

Reply:
Yes. I think I should resolve it before formal pr.

Reply:
Thanks @FrozenGene :+1: